grpcbp/client_middlewares_test.go

```diff
@@ -0,0 +1,174 @@
+package grpcbp
+
+import (
+	"context"
+	"encoding/json"
+	"net"
+	"testing"
+	"time"
+
+	pb "github.com/grpc-ecosystem/go-grpc-middleware/testing/testproto"
+	"github.com/opentracing/opentracing-go"
+	"github.com/reddit/baseplate.go/mqsend"
+	"github.com/reddit/baseplate.go/tracing"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/test/bufconn"
+)
+
+func TestMonitorInterceptor(t *testing.T) {
+	// create test server
+	l, _ := initServer(t)
+
+	// create test client with monitoring interceptor
+	conn := initClient(t, l, grpc.WithUnaryInterceptor(
+		MonitorInterceptor(
+			MonitorClientArgs{
+				ServiceSlug: "test",
+			},
+		),
+	))
+
+	// instantiate gRPC client
+	ctx, mmq := initServerSpan(t)
+	client := pb.NewTestServiceClient(conn)
+
+	t.Run("span-success", func(t *testing.T) {
+		_, err := client.Ping(ctx, &pb.PingRequest{})
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// drain recorder to validate spans
+		msg, err := drainRecorder(mmq)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		var trace tracing.ZipkinSpan
+		err = json.Unmarshal(msg, &trace)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		got := trace.Name
+		want := "test.Ping"
+		if got != want {
+			t.Errorf("got %s, want: %s", got, want)
+		}
+
+		for _, annotation := range trace.BinaryAnnotations {
+			if annotation.Key == "error" {
+				t.Error("got error span, want: success span")
+				break
+			}
+		}
+	})
+
+	t.Run("span-error", func(t *testing.T) {
+		_, err := client.PingError(ctx, &pb.PingRequest{})
+		if err == nil {
+			t.Error("got no error, want error")
+		}
+
+		// drain recorder to validate spans
+		msg, err := drainRecorder(mmq)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		var trace tracing.ZipkinSpan
+		err = json.Unmarshal(msg, &trace)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		got := trace.Name
+		want := "test.PingError"
+		if got != want {
+			t.Errorf("got %s, want: %s", got, want)
+		}
+
+		for _, annotation := range trace.BinaryAnnotations {
+			if annotation.Key == "error" {
+				return
+			}
+		}
+		t.Error("got success span, want: error span")
+	})
+}
+
+func initServer(t *testing.T, opts ...grpc.ServerOption) (*bufconn.Listener, *mockService) {
+	t.Helper()
+
+	// create in-memory connection to reproduce network behavior but without
+	// the need to clean up OS-level resources
```

Comment:
🔕 I don't understand the necessity of this.  Loopback connections are highly reliable and can catch very real bugs.  I strongly encourage using them.  Especially where buffering and timeouts are concerned, it seems far safer to use the OS primitives than trust that the logic in a test-only helper is accurate enough.

Reply:
+1 on having end-to-end tests that test the real thing. Though I will say that in my experience I do see a good bunch of sporadic port not available errors on CIs when picking an available port via `:0`. They also make the tests a bit slower overall.

Reply:
failure to listen on `:0` means there's no port available at all on the machine, I think? that doesn't sound like something that should happen at all (if the machine exhausted all ports it's already in a very very bad state). are you sure it's listing on `:0` failed, not some hardcoded port?

they _might_ be slower, but that's negligible.